lib/experimental/blocks.php

```diff
@@ -83,118 +83,90 @@ function wp_enqueue_block_view_script( $block_name, $args ) {
+		/**
+		 * Process the block bindings attribute.
+		 *
+		 * @param string   $block_content Block Content.
+		 * @param array    $block Block attributes.
+		 * @param WP_Block $block_instance The block instance.
+		 */
+		function process_block_bindings( $block_content, $block, $block_instance ) {
+			// If the block doesn't have the bindings property, return.
+			if ( ! isset( $block['attrs']['metadata']['bindings'] ) ) {
+				return $block_content;
 			}
 
-			if ( 'pattern_attributes' === $attribute_value['source'] ) {
-				if ( ! _wp_array_get( $block_instance->attributes, array( 'metadata', 'id' ), false ) ) {
+			// Assuming the following format for the bindings property of the "metadata" attribute:
+			//
+			// "bindings": {
+			//   "title": {
+			//     "source": {
+			//       "name": "post_meta",
+			//       "attributes": { "value": "text_custom_field" }
+			//     }
+			//   },
+			//   "url": {
+			//     "source": {
+			//       "name": "post_meta",
+			//       "attributes": { "value": "text_custom_field" }
+			//     }
+			//   }
+			// }
+			//
+			global $block_bindings_allowed_blocks;
+			global $block_bindings_sources;
+			$modified_block_content = $block_content;
+			foreach ( $block['attrs']['metadata']['bindings'] as $binding_attribute => $binding_source ) {
+				// If the block is not in the list, stop processing.
+				if ( ! isset( $block_bindings_allowed_blocks[ $block['blockName'] ] ) ) {
+					return $block_content;
+				}
+				// If the attribute is not in the list, process next attribute.
+				if ( ! in_array( $binding_attribute, $block_bindings_allowed_blocks[ $block['blockName'] ], true ) ) {
 					continue;
 				}
-
-				$custom_value = $connection_sources[ $attribute_value['source'] ]( $block_instance, $attribute_name );
-			} else {
-				// If the attribute does not specify the name of the custom field, skip it.
-				if ( ! isset( $attribute_value['value'] ) ) {
+				// If no source is provided, or that source is not registered, process next attribute.
+				if ( ! isset( $binding_source['source'] ) || ! isset( $binding_source['source']['name'] ) || ! isset( $block_bindings_sources[ $binding_source['source']['name'] ] ) ) {
 					continue;
 				}
 
-				// Get the content from the connection source.
-				$custom_value = $connection_sources[ $attribute_value['source'] ](
-					$block_instance,
-					$attribute_value['value']
-				);
-			}
-
-			if ( false === $custom_value ) {
-				continue;
-			}
+				$source_callback = $block_bindings_sources[ $binding_source['source']['name'] ]['apply'];
+				// Get the value based on the source.
+				if ( ! isset( $binding_source['source']['attributes'] ) ) {
+					$source_args = array();
+				} else {
+					$source_args = $binding_source['source']['attributes'];
+				}
+				$source_value = $source_callback( $source_args, $block_instance, $binding_attribute );
+				// If the value is null, process next attribute.
+				if ( is_null( $source_value ) ) {
```

Comment:
I was just reminded that if the attribute accepts `null` or `undefined` as valid values for different meanings, then we should still be able to apply overrides here. Do you think we need a special value here to represent `none`? 🤔

Reply:
@kevin940726 can you elaborate?
What do you think @SantosGuillamot @michalczaplinski?

I wonder how we would handle a `null` or `undefined` value as valid [here](https://github.com/WordPress/gutenberg/blob/a99575c2ea44bc5c5fd172e755f52c16b945d299/lib/experimental/block-bindings/class-wp-block-bindings.php#L139) or [here](https://github.com/WordPress/gutenberg/blob/a99575c2ea44bc5c5fd172e755f52c16b945d299/lib/experimental/block-bindings/class-wp-block-bindings.php#L104). It seems to me that what makes sense intuitively is for `null` or `undefined` to mean that there is no valid value to apply, but I also don't know enough about how developers would use this in the wild to have a strong opinion either way.

Reply:
For context, I was trying to add support for the `linkTarget` attribute of the button block. It accepts a string (like `_blank`) or `undefined` for no target. Let's say we have a source pattern with a button block and its `linkTarget` is `_blank`. Now we want to override the source pattern in an instance to remove the `linkTarget`. We have two problems here:

1. On the editor side, we need a way to represent `undefined` because it's not JSON serializable.
2. On the backend, we need a way to distinguish `NULL` from "skipping the attribute" because sometimes `NULL` is a valid value for some attributes.

I opened two draft PRs exploring different solutions for this problem and I'd love all your feedback!

- #57947: Treats non-existent attributes the same as `null` and uses `null` throughout to represent the attribute.
- #57993: Assumes `undefined` and `null` could be handled differently, and uses a patch similar to JSON patch to represent overrides. This also opens up future possibilities to represent "inserting" a partial override.

Note that the latter changes the `overrides` attribute format so it'd be better if we decide the format sooner than later to save us from back-compat pain 😅.